### DISCUSSION TITLE
Equalizer

### DISCUSSION THREAD

**Discussion by normanalie** [CarlosDerSeher/snapclient#48] [General]
Hello, currently the EQ web interface display 3 bands (Bass, mids and highs) but the code only takes two bands into account.
Is there a reason for that ? DSP capabilities?

**Comment by CarlosDerSeher** [comment]
The EQ isn't quite finished yet. I didn't find time to do it. I'd like to make all those EQ / DSP capabilities accessibly from the web interface at some point though. Maybe you can implement needed functionality yourself?

**Reply by normanalie** [reply]
Yeah, I’ve try to add the code for the mids band and frequency selector on the web interface and its work well on ESP32 WROOM, and I’m gonna add a filter type and Q selector.
My question is now about storage of EQ settings to load them on boot, I’m not very familiar with with esp-idf but from what I’ve understood we can use SPIFFS storage. This partition is used by the HTTP server to store the webpages. Can I use the same partition and store a file with the settings? Should I create a new component to handle file storage for HTTP server and DSP processor ?

**Reply by CarlosDerSeher** [reply]
I would use NVS flash. Have a look at [IDF doc](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/storage/nvs_flash.html)

Hopefully NVS parition will be big enough to store all the data you need. If not we would need to adjust sizes in [partition.csv](https://github.com/CarlosDerSeher/snapclient/blob/master/partitions.csv)

**Reply by normanalie** [reply]
Thanks. NVS is big enough to store filter params. 
I've been playing with filter types, and I don't understand why there is no gain parameter for the peak filter ?
`esp_err_t dsps_biquad_gen_peakingEQ_f32(float *coeffs, float f, float qFactor);`
I think I missed something.

**Reply by CarlosDerSeher** [reply]
I'd expect a gain factor too. Some info about [digital filters](https://webaudio.github.io/Audio-EQ-Cookbook/audio-eq-cookbook.html) can be found here. 

Not sure how to proceed here maybe ask on espressif github page?

**Reply by CarlosDerSeher** [reply]
Not sure how to read the [implementation](https://github.com/espressif/esp-dsp/blob/master/modules/iir/biquad/dsps_biquad_gen_f32.c). It seems gain is set to 1 but still the formulas don't add up. Which IIR form did they implement?

**Reply by CarlosDerSeher** [reply]
Ok probably more research on dsp needs to ne done. See [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.iirpeak.html) for example. Quality factor  is related to -3dB bandwidth of the filter.
